Record `flutter run` success/fail, build mode, platform, start time in analytics

xster opened the pull request:
- Create a generic FlutterCommandResult plumbing for FlutterCommand subclasses to pipe back analytics messages
- Instead of recording the entire resident runner session, record the app speed of `flutter run` to the app start
- Record `flutter run` success/fail, build mode, hot/cold, platform, emulator.

Fixes #9500 
Fixes #9499
Fixes #9498

xster committed:
Command infra

xster committed:
done logic

xster committed:
tested

xster committed:
rename to FlutterCommandResult

xster committed:
Add some comments

xster committed:
Add platform, emulator too

xster committed:
Add tests

xster commented:
Added a clock to the testing context

tvolkert commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -111,18 +144,35 @@ abstract class FlutterCommand extends Command<Null> {
+    final DateTime endTime = clock.now();
+    printTrace("'flutter $name' took ${endTime.difference(startTime).inMilliseconds}ms.");
+    if (usagePath != null) {
+      final String label = (
```
The following phrasing is a little bit more plain, and it allow you to get rid of `_getNullableExitCodeName()` altogether.

```
final List<String> labels = <String>[];
if (commandResult?.exitCode != null)
  labels.add(getEnumName(commandResult.exitCode));
if (commandResult?.analyticsParameters?.isNotEmpty ?? false)
  labels.addAll(commandResult.analyticsParameters);
final String label = labels.join('-');
```

xster replied:
Thanks for the tweak.

tvolkert commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,38 @@ import 'flutter_command_runner.dart';
+  /// latency without measuring user interaction time. 
+  /// 
+  /// Will automatically measure the command's complete time if not provided. 
+  DateTime exitTime;
```
Is there a use case for commands providing their own entry time as well?

xster replied:
No. I think we should measure user perceivable time. That latency measurement should start as early as possible.

tvolkert commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -111,18 +144,35 @@ abstract class FlutterCommand extends Command<Null> {
-    final Stopwatch stopwatch = new Stopwatch()..start();
-    final UsageTimer analyticsTimer = usagePath == null ? null : flutterUsage.startTimer(name);
+  Future<Null> run() async {
+    final DateTime startTime = clock.now();
```
By starting this here, we'll record the time it takes to update the cache if needed (which is done in `verifyThenRunCommand()`), which will mess with the analytics results...

xster replied:
Ditto. I think that's fine as a metric to optimize for.

tvolkert commented on packages/flutter_tools/lib/src/usage.dart:
```diff
@@ -80,8 +80,19 @@ class Usage {
+    );
   }
 
   UsageTimer startTimer(String event) {
```
Is this used anymore?  It looks like `UsageTimer` can be nuked.

xster replied:
Done. Love deleting code

tvolkert commented on packages/flutter_tools/test/src/context.dart:
```diff
@@ -55,7 +56,8 @@ void _defaultInitializeContext(AppContext testContext) {
     ..putIfAbsent(Usage, () => new MockUsage())
-    ..putIfAbsent(FlutterVersion, () => new MockFlutterVersion());
+    ..putIfAbsent(FlutterVersion, () => new MockFlutterVersion())
+    ..putIfAbsent(Clock, () => new Clock());
```
Do you mean `MockClock` (or `FakeClock` given my comment below)?

xster replied:
Ah ya. I was gonna switch everything without breaking the existing tests but there were too many. Done. We can swap the others as needed.

tvolkert commented on packages/flutter_tools/test/src/context.dart:
```diff
@@ -234,3 +236,5 @@ class _MockUsageTimer implements UsageTimer {
 
 class MockFlutterVersion extends Mock implements FlutterVersion {}
+
+class MockClock extends Mock implements Clock {}
```
It's such a simple interface, it may be worth writing a `FakeClock` and saving your tests the effort of mocking...  It could have a setter for setting the current time, and it'd increment the value every time `now()` was called.

xster replied:
I think it's ok for the test to be verbose wrt the input/output near the assertions.

tvolkert reviewed (changes requested):
Travis and AppVeyor seem unhappy.

xster committed:
review notes

xster commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -288,11 +292,30 @@ class RunCommand extends RunCommandBase {
+    //
+    // Do not add more operations to the future.
+    final Completer<Null> appStartedTimeRecorder = new Completer<Null>.sync()
+        ..future.then((Null _) { appStartedTime = clock.now(); });
```
Done

xster commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,38 @@ import 'flutter_command_runner.dart';
 
 typedef void Validator();
 
+enum ExitCode {
```
Done

xster commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,38 @@ import 'flutter_command_runner.dart';
+  return code == null ? null : getEnumName(code);
+}
+
+class FlutterCommandResult {
```
Done

xster commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,38 @@ import 'flutter_command_runner.dart';
+
+class FlutterCommandResult {
+  FlutterCommandResult(
+    this.exitCode,
```
Ok

xster commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,38 @@ import 'flutter_command_runner.dart';
+    }
+  ) { assert(exitCode != null); }
+
+  ExitCode exitCode;
```
Done

xster commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,38 @@ import 'flutter_command_runner.dart';
+  ) { assert(exitCode != null); }
+
+  ExitCode exitCode;
+  /// Optional dimension data that can be appended to the timing event.
```
Done

xster commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,38 @@ import 'flutter_command_runner.dart';
+  /// complete during the command's execution. Use to measure user perceivable
+  /// latency without measuring user interaction time. 
+  /// 
+  /// Will automatically measure the command's complete time if not provided. 
```
Done

xster committed:
default to MockClock

devoncarew commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -22,6 +23,40 @@ import 'flutter_command_runner.dart';
+  /// 
+  /// [FlutterCommand] will automatically measure and report the command's
+  /// complete time if not provided.
+  final DateTime exitTime;
```
`endTime`? `endTimeOverride`?

xster replied:
Sure. Renamed to endTimeOverride

devoncarew commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -243,10 +242,15 @@ class RunCommand extends RunCommandBase {
-      return null;
+      return new FlutterCommandResult(
+        ExitStatus.success,
+        analyticsParameters: <String>["daemon"],
```
nit: I think flutter style would want single quotes here

xster replied:
ah thanks

devoncarew commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -111,18 +146,37 @@ abstract class FlutterCommand extends Command<Null> {
+    final FlutterCommandResult commandResult = await verifyThenRunCommand();
+
+    final DateTime endTime = clock.now();
+    printTrace("'flutter $name' took ${endTime.difference(startTime).inMilliseconds}ms.");
```
There's a utility to print ms - `getElapsedAsMilliseconds(Duration)`. Generally used for user facing output (non-trace output), but perhap useful here.

xster replied:
oh cool, thanks

devoncarew commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -148,11 +202,13 @@ abstract class FlutterCommand extends Command<Null> {
       flutterUsage.sendCommand(usagePath);
 
-    await runCommand();
+    return await runCommand();
```
you don't need the await here, since you're now returning the Future

tvolkert replied:
I've heard that while the VM unwraps futures in these cases, it's bad style, but then I've also heard it's fiiiine.  @abarth or @Hixie, is there a definitive answer?

xster replied:
I think semantically, we don't need to await here

tvolkert commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -288,11 +292,32 @@ class RunCommand extends RunCommandBase {
+    // Do not add more operations to the future.
+    final Completer<Null> appStartedTimeRecorder = new Completer<Null>.sync();
+    appStartedTimeRecorder.future.then(
+      (Null _) { appStartedTime = clock.now(); }
```
Nit: for `_` parameters, you can omit the type.

Hixie replied:
I'm confused as to what this is trying to do. Wouldn't `scheduleMicrotask` do the same thing cleanly? Or `Future.sync`?

xster replied:
oh cool, thanks

xster replied:
oops, our comments crossed paths midair. 

I didn't really understand the references to Future in the Completer's dartdoc either. It seems to be written for the perspective of the Completer's reader rather than the Completer's supplier. 

I'm trying to supply a Completer such that a piece of code runs synchronously when the owner of the Completer, a few layers deeper, calls complete() on it.

Hixie replied:
I'm not sure what "supply a completer" means. Normally a completer is private to the person who creates it, and you hand out its future.

tvolkert replied:
@Hixie I had the same reaction, but it looks like it's an existing API in tools.  I plan to remedy this in a separate PR by having the caller pass a callback instead of a completer.

xster replied:
cool neat. It doesn't actually seem heavily used in the codebase. A simpler callback would indeed be easier to parse.

tvolkert commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -288,11 +292,32 @@ class RunCommand extends RunCommandBase {
+      analyticsParameters: <String>[
+        hotMode ? 'hot' : 'cold',
+        getModeName(getBuildMode()),
+        getNameForTargetPlatform(device.targetPlatform),
```
If you're synced to tip-of-tree, this now needs to be `await device.targetPlatform`

xster replied:
thanks

tvolkert commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -288,11 +292,32 @@ class RunCommand extends RunCommandBase {
+        hotMode ? 'hot' : 'cold',
+        getModeName(getBuildMode()),
+        getNameForTargetPlatform(device.targetPlatform),
+        device.isLocalEmulator ? 'emulator' : null,
```
Likewise, `await device.isLocalEmulator`

xster replied:
done

tvolkert commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -111,18 +146,37 @@ abstract class FlutterCommand extends Command<Null> {
+      if (commandResult?.exitStatus != null)
+        labels.add(getEnumName(commandResult.exitStatus));
+      if (commandResult?.analyticsParameters?.isNotEmpty ?? false)
+        labels.addAll(commandResult?.analyticsParameters);
```
No need for the `?` here

xster replied:
done

tvolkert commented on packages/flutter_tools/lib/src/usage.dart:
```diff
@@ -80,15 +80,19 @@ class Usage {
+    String category, 
+    String variableName, 
+    Duration duration,
+    {
```
Similar formatting comment here:

```
  Duration duration, {
  String label,
```

xster replied:
done

tvolkert commented on packages/flutter_tools/lib/src/usage.dart:
```diff
@@ -80,15 +80,19 @@ class Usage {
+    {
+      String label,
+    }) {
+    _analytics.sendTiming(
```
We used to check for `suppressAnalytics`...

xster replied:
nice catch, thanks!

tvolkert commented on packages/flutter_tools/test/src/runner/flutter_command_test.dart:
```diff
@@ -38,12 +44,93 @@ void main() {
+      );
+    },
+    overrides: <Type, Generator>{
+      Clock: () => clock,
```
`MockUsage` and `MockClock` are seeded in the context for tests by default, so you don't need to instantiate and override them here. (same comment in the tests below)

xster replied:
so I ended up defaulting back to a real clock. I think it's more convenient as we move more and more stuff to Clock for tests that don't really care to keep working without null handling and not get an arbitrary value.

xster committed:
fix and use quiver

xster commented:
Ah, I was wondering why travis was failing. Rebased. 

Also TIL quiver. Removed Clock since there already is one and used some of the string functions

xster committed:
review comments

xster committed:
not sure why travis is failing but tests are not failing locally

xster committed:
add one more test, fix for travis

tvolkert commented on packages/flutter_tools/lib/src/usage.dart:
```diff
@@ -81,15 +81,20 @@ class Usage {
+    String label,
+    }) {
+    if (!suppressAnalytics) {
+    _analytics.sendTiming(
```
Nit: indent

xster committed:
nit fix

xster committed:
Merge branch 'master' into analytics

xster committed:
handle multi-device reporting
